### PR TITLE
experimental proposal DCP v2

### PR DESCRIPTION
**Pull request by teja-rao**
Just for fun - hacking up a different way to do checkpointing. Dont take this too seriously, it may not ever see light!

This reimagines distributed checkpoints - preserving torch serialization format, files per rank and goes for heavy simplification of implementation. 

 The following concepts are eliminated - 
- stateful protocol which is a source of issues like https://github.com/pytorch/pytorch/issues/146157, cause of stuck jobs because of complexity in handling failures in collectives in state_dict() calls along calls in dcp, complexities of managing current cuda devices, backward compatibility issues.
- planners which are a confusing concept to map state_dict to internal representation of storage. planners also build metadata, allow advanced write strategies which can all be done much simpler independently. 
- Internal structures like WriteItem, ReadItem and all the code to translate a param to storage chunk (which is redundant ) to improve debuggability.
- do not require an additional pg to be constructed which is critical problem for large jobs due to gloo initialization times.

The following are introduced/modified - 
- Standardized on torch serialization format which I know will be appreciated by ML engineers and hopefully increase dcp adoption in research.
- Introduced a stateful checkpointer class to manage resources properly for async checkpointing.
- Eliminated collectives in save and load path (assuming metadata can be cached). 
- Decouples async checkpointing specifically staging from StoragerWriter abstraction enabling more code reuse.
- Abstraction to control layouts and serialization format with common usecases implemented.
- Introduced an optional barrier to allow waiting for all ranks to finish checkpointing which is useful for async checkpointing.

TBD
- support streaming serialization/deserialization (and streaming support in torch serialization?)
- storage/APIs need a second look.
- Support in torch serialization to load a specific storage without loading entire model file (This can be done without format changes).
- Additional components that can help with easier integration like checkpoint scheduler, factory methods for creation, config management .
- need to think through validations to catch common pitfalls when using the apis.
- implement support filtering replicated tensors. 

to review: start with _base.py and _checkpointer.py and move on to _checkpoint_loader.py.

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @LucasLLC @MeetVadakkanchery @mhorowitz @pradeepfn @ekr0

### CODE REVIEW

File: torch/distributed/checkpoint/_v2/_checkpointing.py [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+            context (CheckpointContext): The context to save the checkpoint.
+            root_dir (str): The path to save the checkpoint.
+            use_cached_manifest (bool): Whether to use cached manifest. (Default: ``False``). If
+                use_cached_manifest is True, it will use the compute the manifest to save the checkpoint.
+                Otherwise, it will compute the manifest and save the checkpoint. If the checkpoint is saved
+                successfully, it will return a tuple of two futures. The first future is a future for the
+                manifest and the second future is a future for the checkpoint.
+
+        Returns:
+            tuple[Future, Future]: A tuple of two futures. The first future can be awaited on for completion
```
Comment (wconstab) [inline torch/distributed/checkpoint/_v2/_checkpointing.py:111]: conflicting definition?

above:
" The first future is a future for the
                manifest and the second future is a future for the checkpoint."

File: torch/distributed/checkpoint/_v2/_checkpointing.py [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+                "optimizer": optimizer_state_dict,
+                "dataloader": dataloader_state_dict,
+                "metrics": metrics_state_dict,
+                "extra_state": extra_state_dict,
+            }
+
+        We expect most users to use a variant of async checkpointing implementation and so the API is
+        designed to be easy to use for this case. A synchronous checkpointing implementation
+        can still be implemented by on top of this API. EG: synchronous version can return a tuple of
+        two futures which are set with a result in the save() method.
```
Comment (wconstab) [inline torch/distributed/checkpoint/_v2/_checkpointing.py:98]: yea, though it might be more convenient if the sync version returns None (though users are free to ignore the futures as well)

File: torch/distributed/checkpoint/_v2/_checkpointing.py [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+    @abc.abstractmethod
+    def save(
+        self,
+        state_dict: dict[str, Any],
+        context: CheckpointContext,
+        root_dir: str,
+        use_cached_manifest: bool = False,
+    ) -> Optional[tuple[Future[None], Future[None]]]:
+        """
+        Save a checkpoint to a given path on storage. This optionally saves the metadata aggregated
```
Comment (wconstab) [inline torch/distributed/checkpoint/_v2/_checkpointing.py:81]: what determines the saving of metadata? 

if false, does it mean no metadata is saved, or it just means only the local metadata is saved?  I assume the latter.  Then it is still possible but less efficient to work with the files later.

File: torch/distributed/checkpoint/_v2/_checkpointing.py [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+
+
+class ManifestBuilder(abc.ABC):
+    """
+    This class is responsible for building the manifest for a checkpoint.
+
+    TODO alernate plan:
+    We can also avoid this class by just providing a utility function to build the manifest from
+    ShardedTensor and DTensor. If the user has a state_dict with custom impls of parallelisms, they
+    can write their own logic. User can call the function before saving the checkpoint and pass the
```
Comment (wconstab) [inline torch/distributed/checkpoint/_v2/_checkpointing.py:380]: it would be interesting to mock up what the api a user would call to specify their own manifest would look like.

File: torch/distributed/checkpoint/_checkpointer.py [hunk 5]
```diff
@@ -1,100 +0,0 @@
...
-    ):
-        """Initializes the Checkpointer instance.
-
-        Args:
-            storage_writer: Instance of StorageWrite use to perform writes.
-            storage_reader: StorageReader used to load data from.
-            process_group: ProcessGroup to be used for cross-rank synchronization.
-            coordinator_rank: Rank to use to coordinate the checkpoint. rank0 is used by default.
-            no_dist: If ``True``, distributed checkpoint will not load in SPMD style. (Default: ``False``)
-            loader_planner: Instance of LoadPlanner to use when loading.
```
Comment (wconstab) [inline torch/distributed/checkpoint/_checkpointer.py:51]: by removing the planner in the below design, which of the new components takes over its responsibility?  iiuc this responsibility is mainly determining which fqns and shards will be saved / loaded by each rank?

are we just redistributing this same functionality, or are we also simplifying the functionality (e.g we are giving up on some optimizations on deduplication, or loading once and broadcasting over nccl vs loading multiple times?)

File: torch/distributed/checkpoint/_v2/_checkpointing.py (hunk 2 above)
Comment (teja-rao) [reply torch/distributed/checkpoint/_v2/_checkpointing.py:98]: yeah, i was thinking about this. I was expecting almost everyone to use async implementation - so i tried to return futures to avoid branches but I can change this to Optional[tuple[Future,Future]] to make it semantically correct. Probably makes it easy for users to understand the expectations.

File: torch/distributed/checkpoint/_v2/_checkpointing.py (hunk 3 above)
Comment (teja-rao) [reply torch/distributed/checkpoint/_v2/_checkpointing.py:81]: 1. this will be a config option that user explicitly sets. We cannot support resharding or filter replicated tensors if users disabled manifest savings. We validate options to make sure to ensure compatibility. EG: user cannot enable replicated tensors and disable manifest saving. 

One clarification on terminology - I am defining two structures here - manifest: this contains all state_dict metadata across all ranks and  Metadata: this is an encompassing structure that contains manifest and other static metadata like version or config information that might be needed for backward compatibility when we change implementations. We would always write the metadata now but avoid building manifest.  

2. If we see value in having local manifest data - we can do that. This data is already in data.pkl in serialized file (from torch.save). One advantage from adding local manifest is that it makes debugging easy. I think we should still expect all serialization formats to have this data.

File: torch/distributed/checkpoint/_v2/_checkpointing.py (hunk 4 above)
Comment (teja-rao) [reply torch/distributed/checkpoint/_v2/_checkpointing.py:380]: thanks! will add this.

File: torch/distributed/checkpoint/_checkpointer.py (hunk 5 above)
Comment (teja-rao) [reply torch/distributed/checkpoint/_checkpointer.py:51]: We we are mostly redistributing the parts of planner - which are responsible for - building global metadata, filtering replicated tensors and creating write plan. We avoid building the write plan and just assume we create a new state_dict and pass it down after filtering replicated tensors if needed. Passing around state_dict makes it easy to debug than having to understand the plan structs. 

Breaking planner into smaller parts also helps with readability and extensibility. We dont need to implement sub-classes for planner to customize smaller parts.

File: torch/distributed/checkpoint/_v2/_checkpoint_loader.py [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+    """
+    TorchStreamingLoader or TorchPartialLoader will be implemented outside of checkpointing
+    and almost certainly with different API. This is quick hack to show what we need
+    from torch load to support checkpointing usecases: resharding, low memory footprint,
+    random access to params, etc.
+
+    The capabilities we need -
+        1. Load a specified param without loading the entire model file into memory.
+        2. Load specified params to a specified pre-allocated location in chunks or return
+            a generator so caller can load the params in chunks.
```
Comment (mikaylagawarecki) [inline torch/distributed/checkpoint/_v2/_checkpoint_loader.py:43]: A couple thoughts here:

(1) Given `torch.save`-ed checkpoint, get metadata (e.g. shapes, strides, tensor types about what is in the checkpoint). For this loading under a FakeTensorMode would be helpful (this would not read any actual data bytes)


```python
import torch
import torch.nn as nn
from torch._subclasses.fake_tensor import FakeTensorMode


m = nn.Linear(2, 3)
torch.save(m.state_dict(), "foo.pt")


with FakeTensorMode():
    sd = torch.load("foo.pt")

print(sd)
OrderedDict([('weight', FakeTensor(..., size=(3, 2))), ('bias', FakeTensor(..., size=(3,)))])
```

From this you can get various metadata you need e.g.

```python
>>> sd['weight'].untyped_storage().nbytes()
96
```

(2) Another thing that you would want to know is the offset of a tensor's **storage** within the checkpoint 

For this, with a patch like https://github.com/pytorch/pytorch/pull/147787, you would have

```python
>>> sd.weight.untyped_storage()._checkpoint_offset
832
```

(if the checkpoint contains wrapper tensor subclasses like DTensors you would have to access the untyped_storage() of the plain tensor being wrapped)

Note: the current implementation of getting the offset might trigger many reads over a network (as it reads the end of central directory record to determine the offset, but that can be improved (if that is the usecase) 

I think this should give enough information to do what's described. As to the loading in chunks part, is it accurate to say that one the offsets are known that could be handled separately by writing a custom reader?
